File: apps/mobile/ios/App/App/PerformanceDiagnosticsPlugin.swift
import Capacitor
import Foundation
import OSLog

@objc(PerformanceDiagnosticsPlugin)
final class PerformanceDiagnosticsPlugin: CAPPlugin, CAPBridgedPlugin {
    let identifier = "PerformanceDiagnosticsPlugin"
    let jsName = "PerformanceDiagnostics"
    let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "logTiming", returnType: CAPPluginReturnPromise)
    ]

    private let logger = Logger(
        subsystem: Bundle.main.bundleIdentifier ?? "org.speleodb.app",
        category: "SpeleoDBPerformance"
    )

    @objc func logTiming(_ call: CAPPluginCall) {
        guard let scope = call.getString("scope"),
              let runId = call.getInt("runId"),
              let phase = call.getString("phase"),
              let status = call.getString("status"),
              let message = PerformanceTimingLogFormatter.format(
                scope: scope,
                runId: runId,
                phase: phase,
                durationMs: call.getDouble("durationMs"),
                status: status
              ) else {
            call.reject("Invalid performance timing record", "E_PERFORMANCE_TIMING")
            return
        }

        logger.notice("\(message, privacy: .public)")
        call.resolve()
    }
}

File: apps/mobile/android/app/src/main/java/org/speleodb/app/PerformanceDiagnosticsPlugin.java
package org.speleodb.app;

import android.util.Log;

import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;

@CapacitorPlugin(name = "PerformanceDiagnostics")
public final class PerformanceDiagnosticsPlugin extends Plugin {
    private static final String ERROR_CODE = "E_PERFORMANCE_TIMING";
    private static final String LOG_TAG = "SpeleoDBPerformance";

    @PluginMethod
    public void logTiming(PluginCall call) {
        String scope = call.getString("scope");
        Integer runId = call.getInt("runId");
        String phase = call.getString("phase");
        Double durationMs = call.getDouble("durationMs");
        String status = call.getString("status");

        String message = runId == null
            ? null
            : PerformanceTimingLogFormatter.format(scope, runId, phase, durationMs, status);
        if (message == null) {
            call.reject("Invalid performance timing record", ERROR_CODE);
            return;
        }

        Log.i(LOG_TAG, message);
        call.resolve();
    }
}
